Jenkins fails to boot due to wrong ruby version

TemplateID: jenkins/jenkins2-ha-agents.yaml
Region: eu-west-1a

*Jenkins fails to boot* because:
- gem fails to install `aws-sdk-autoscaling` (version 1.47.0) and `aws-sdk-sqs` (version 1.34.0) because
- they depend on `aws-partitions` who was released with a non-breaking SemVer (1.494.0 to 1.496.0) when it's actually breaking vs 1.493.0 (requiring ruby >= 2.3 when 1.493.0 didn't)

Possible fix is to force installation of ruby 2.4 instead of ruby 2 by:
```diff
--- a/jenkins/jenkins2-ha-agents.yaml
+++ b/jenkins/jenkins2-ha-agents.yaml
@@ -968,8 +968,6 @@ Resources:
           packages:
             rpm:
               jenkins: 'http://ftp-chi.osuosl.org/pub/jenkins/redhat-stable/jenkins-2.303.1-1.1.noarch.rpm'
-            yum:
-              'ruby': []
             rubygems:
               'aws-sdk-autoscaling': ['1.47.0']
               'aws-sdk-sqs': ['1.34.0']
@@ -1376,6 +1374,7 @@ Resources:
         UserData:
           'Fn::Base64': !Sub |
             #!/bin/bash -x
+            amazon-linux-extras install ruby2.4
             /opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource MasterLaunchTemplate --region ${AWS::Region}
             /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource MasterASG --region ${AWS::Region}
   MasterASG:
@@ -1866,8 +1865,6 @@ Resources:
                 ensureRunning: !If [HasSystemsManagerAccess, true, false]
         setup:
           packages:
-            yum:
-              'ruby': []
             rubygems:
               'aws-sdk-autoscaling': ['1.47.0']
               'aws-sdk-ec2': ['1.200.0']
@@ -2099,6 +2096,7 @@ Resources:
         UserData:
           'Fn::Base64': !Sub |
             #!/bin/bash -x
+            amazon-linux-extras install ruby2.4
             /opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource AgentLaunchTemplate --region ${AWS::Region}
             /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource AgentASG --region ${AWS::Region}
   AgentASG:
```

I was working on the same issue today.  Apparently, the issue is recent ruby modules are requiring ruby >= 2.3   Even though the modules in the template are pinned at certain versions, it seems some rub modules may be allowed to link to the latest for their dependencies.  For my version of the Jenkins CF template the issue was specifically with aws-eventstream v1.2.0 requiring ruby >= 2.3

Some of my notes are below.  I'll give your solution a try with my template, thanks for figuring it out!
```
/var/log/cloud-init-output.log
Cloud-init v. 19.3-3.amzn2 running 'modules:final' at Fri, 03 Sep 2021 22:12:51 +0000. Up 32.12 seconds.
+ /opt/aws/bin/cfn-init -v --stack Jenkins --resource AgentLC --region us-east-2
Error occurred during build: Failed to install gem: aws-sdk-ec2-1.62.0 (return code 1)
+ /opt/aws/bin/cfn-signal -e 1 --stack Jenkins --resource AgentASG --region us-east-2
ValidationError: Stack arn:aws:cloudformation:us-east-2:815262906617:stack/Jenkins/7daca480-efca-11ea-be61-0aa874bc8df2 is in UPDATE_COMPLETE state and cannot be signaled



/var/log/cfn-init.log
021-09-03 21:55:59,220 [ERROR] Gem failed. Output: ERROR:  Error installing aws-sdk-ec2:
        aws-eventstream requires Ruby version >= 2.3.

021-09-03 21:55:59,220 [ERROR] Error encountered during build of setup: Failed to install gem: aws-sdk-ec2-1.62.0 (return code 1)
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/construction.py", line 542, in run_config
    CloudFormationCarpenter(config, self._auth_config).build(worklog)
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/construction.py", line 229, in build
Hangupanges['packages'][manager] = CloudFormationCarpenter._packageTools[manager]().apply(packages, self._auth_config)
sh-4.2$ ls -lapython2.7/site-packages/cfnbootstrap/lang_package_tools.py", line 111, in apply
    if self._install_gem(pkg, ver):
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/lang_package_tools.py", line 155, in _install_gem
    raise ToolError("Failed to install gem: %s-%s" % (pkg, ver), result.returncode)
ToolError: Failed to install gem: aws-sdk-ec2-1.62.0 (return code 1)
2021-09-03 21:55:59,222 [ERROR] -----------------------BUILD FAILED!------------------------
2021-09-03 21:55:59,222 [ERROR] Unhandled exception during build: Failed to install gem: aws-sdk-ec2-1.62.0 (return code 1)
Traceback (most recent call last):
  File "/opt/aws/bin/cfn-init", line 171, in <module>
    worklog.build(metadata, configSets)
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/construction.py", line 129, in build
    Contractor(metadata).build(configSets, self)
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/construction.py", line 530, in build
    self.run_config(config, worklog)
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/construction.py", line 542, in run_config
    CloudFormationCarpenter(config, self._auth_config).build(worklog)
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/construction.py", line 229, in build
    changes['packages'][manager] = CloudFormationCarpenter._packageTools[manager]().apply(packages, self._auth_config)
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/lang_package_tools.py", line 111, in apply
    if self._install_gem(pkg, ver):
  File "/usr/lib/python2.7/site-packages/cfnbootstrap/lang_package_tools.py", line 155, in _install_gem
    raise ToolError("Failed to install gem: %s-%s" % (pkg, ver), result.returncode)
ToolError: Failed to install gem: aws-sdk-ec2-1.62.0 (return code 1)

Agents must use Ruby 2.0
root@ip-10-0-37-179 bin]# yum list available ruby
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Available Packages
ruby.x86_64               

AWS Eventstream adds:
>  spec.required_ruby_version = '>= 2.3'
see here:
https://github.com/aws/aws-sdk-ruby/blob/a6736ea94dc708bfea6a7099dd11d40931a527a1/gems/aws-eventstream/aws-eventstream.gemspec#L18
```

--- a/jenkins/jenkins2-ha-agents.yaml
+++ b/jenkins/jenkins2-ha-agents.yaml
@@ -964,12 +964,16 @@ Resources:
             yum:
               'java-11-amazon-corretto-headless': []
               'daemonize': []
+              ruby: []
+            rubygems: # workaround for https://github.com/aws/aws-sdk-ruby/pull/2573#issuecomment-913984611
+              'aws-eventstream': ['1.1.1']
+              'aws-partitions': ['1.493.0']
+              'aws-sigv4': ['1.2.4']
+              'aws-sdk-core': ['3.119.1']
         install:
           packages:
             rpm:
               jenkins: 'http://ftp-chi.osuosl.org/pub/jenkins/redhat-stable/jenkins-2.303.1-1.1.noarch.rpm'
-            yum:
-              'ruby': []
             rubygems:
               'aws-sdk-autoscaling': ['1.47.0']
               'aws-sdk-sqs': ['1.34.0']
@@ -1826,6 +1830,12 @@ Resources:
           packages:
             yum:
               'java-11-amazon-corretto-headless': []
+              ruby: []
+            rubygems: # workaround for https://github.com/aws/aws-sdk-ruby/pull/2573#issuecomment-913984611
+              'aws-eventstream': ['1.1.1']
+              'aws-partitions': ['1.493.0']
+              'aws-sigv4': ['1.2.4']
+              'aws-sdk-core': ['3.119.1']
           files:
             '/etc/cfn/cfn-hup.conf':
               content: !Sub |
@@ -1866,8 +1876,6 @@ Resources:
                 ensureRunning: !If [HasSystemsManagerAccess, true, false]
         setup:
           packages:
-            yum:
-              'ruby': []
             rubygems:
               'aws-sdk-autoscaling': ['1.47.0']
               'aws-sdk-ec2': ['1.200.0']

